anvil/observables.py

```diff
@@ -3,35 +3,63 @@
 
 functions for calculating observables on a stack of states.
 
-Functions
----------
-print_plot_observables:
-    can be considered the main function which in turn calculates the rest
-    of the observables defined in this module.}
-
 Notes
 -----
 Check the definitions of functions, most are defined according the the arxiv
 version: https://arxiv.org/pdf/1904.12072.pdf
 
 """
-from math import acosh
-
-import pandas as pd
-import matplotlib.pyplot as plt
 import numpy as np
 from scipy.signal import correlate
+import torch
+from tqdm import tqdm
 
-from reportengine.table import table
-from reportengine.figure import figure
+def arcosh(x):
+    """Inverse hyperbolic cosine function for torch.Tensor arguments.
 
+        arcosh(x) = log_e (x + sqrt(x^2 - 1) )
+                  = log_e (x) + log_e (1 + sqrt(x^2 - 1) / x)
+    """
+    c0 = torch.log(x)
+    c1 = torch.log1p(torch.sqrt(x * x - 1) / x)
+    return c0 + c1
 
-class GreenFunction:
-    def __init__(self, states, geometry):
+
+def bootstrap(observable):
+    """Return the standard deviation for an observable based on a number of
+    'bootstrap' samples."""
+    obs_full = observable[0]
+    obs_bootstrap = observable[1:]
+
+    variance = torch.mean((obs_bootstrap - obs_full) ** 2, axis=0)
+    #bias = torch.mean(obs_bootstrap) - obs_full  # not sure whether to use this
+
+    return variance.sqrt()
+
+
+class TwoPointFunction:
+    def __init__(self, states, geometry, bootstrap_n_samples):
         self.geometry = geometry
-        self.sample = states
+        self.states = states
+        self.n_samples = bootstrap_n_samples
+        self.sample_size = states.size(0)
 
-    def __call__(self, x_0: int, x_1: int):
+        # Bootstrap samples of size (n_samples, sample_size, n_states)
+        self.sample_indices = np.random.choice(
```

Comment:
Mutating objects is discouraged in reportengine apps because the philosophy is you have nodes in your graph of dependencies (either things which are parsed/produced in config or functions defined in your providers module) and these things don't get mutated so can be used multiple times.

Now we cheated a little with this class because we instead wanted an object which could be called by other functions without having loads of boilerplate, but the the original method always created a new object in the call method and then returned it.

admittedly you set self.value = [] each time you use this method, however there is a parallel mode to anvil-sample and if one was using this then it could have unintended side effects.